daemon/attach.go

```diff
@@ -113,138 +113,109 @@ func (daemon *Daemon) ContainerAttach(job *engine.Job) engine.Status {
-							cStderr.Close()
-						}
-					}()
-				}
-				if tty {
-					_, err = utils.CopyEscapable(cStdin, stdin)
-				} else {
-					_, err = io.Copy(cStdin, stdin)
-
-				}
-				if err == io.ErrClosedPipe {
-					err = nil
-				}
-				if err != nil {
-					log.Errorf("attach: stdin: %s", err)
-				}
-				errors <- err
-			}()
-		}
+		cStdin = streamConfig.StdinPipe()
 	}
+
 	if stdout != nil {
-		nJobs++
-		// Get a reader end of a pipe that is attached as stdout to the container.
-		if p, err := streamConfig.StdoutPipe(); err != nil {
-			errors <- err
-		} else {
-			cStdout = p
-			go func() {
-				log.Debugf("attach: stdout: begin")
-				defer log.Debugf("attach: stdout: end")
-				// If we are in StdinOnce mode, then close stdin
-				if stdinOnce && stdin != nil {
-					defer stdin.Close()
-				}
-				_, err := io.Copy(stdout, cStdout)
-				if err == io.ErrClosedPipe {
-					err = nil
-				}
-				if err != nil {
-					log.Errorf("attach: stdout: %s", err)
-				}
-				errors <- err
-			}()
+		cStdout = streamConfig.StdoutPipe()
+	}
+
+	if stderr != nil {
+		cStderr = streamConfig.StderrPipe()
+	}
+
+	// Connect stdin of container to the http conn.
+	stdinErr = promise.Go(func() error {
+		if stdin == nil || !openStdin {
+			return nil
 		}
-	} else {
-		// Point stdout of container to a no-op writer.
-		go func() {
-			if cStdout, err := streamConfig.StdoutPipe(); err != nil {
-				log.Errorf("attach: stdout pipe: %s", err)
+		wg.Add(1)
+		log.Debugf("attach: stdin: begin")
+		defer func() {
+			if stdinOnce && !tty {
+				cStdin.Close()
 			} else {
-				io.Copy(&ioutils.NopWriter{}, cStdout)
+				// No matter what, when stdin is closed (io.Copy unblock), close stdout and stderr
+				if cStdout != nil {
+					cStdout.Close()
+				}
+				if cStderr != nil {
+					cStderr.Close()
+				}
 			}
+			wg.Done()
+			log.Debugf("attach: stdin: end")
 		}()
-	}
-	if stderr != nil {
-		nJobs++
-		if p, err := streamConfig.StderrPipe(); err != nil {
-			errors <- err
+		var err error
+		if tty {
+			_, err = utils.CopyEscapable(cStdin, stdin)
 		} else {
-			cStderr = p
-			go func() {
-				log.Debugf("attach: stderr: begin")
-				defer log.Debugf("attach: stderr: end")
-				// If we are in StdinOnce mode, then close stdin
-				// Why are we closing stdin here and above while handling stdout?
-				if stdinOnce && stdin != nil {
-					defer stdin.Close()
-				}
-				_, err := io.Copy(stderr, cStderr)
-				if err == io.ErrClosedPipe {
-					err = nil
-				}
-				if err != nil {
-					log.Errorf("attach: stderr: %s", err)
+			_, err = io.Copy(cStdin, stdin)
+
+		}
+		if err == io.ErrClosedPipe {
+			err = nil
+		}
+		if err != nil {
+			log.Errorf("attach: stdin: %s", err)
+			return err
+		}
+		return nil
+	})
+
+	attachStream := func(name string, stream io.Writer, streamPipe io.ReadCloser) chan error {
+		if stream == nil {
+			chErr := make(chan error, 1)
+			chErr <- nil
+			return chErr
+		}
+		return promise.Go(func() error {
+			wg.Add(1)
```

Comment:
It is better to do `Add` outside goroutine, because race detector will be angry about this. Also I want you back to one channel and simple goroutines back. I'll tell why below.